File: backend/src/main/java/com/smartcampus/service/BookingService.java
package com.smartcampus.service;

import com.smartcampus.dto.BookingDTO;

import java.util.List;

/**
 * ================================================================
 * BookingService — Booking Management Interface
 * ================================================================
 *
 * WORKFLOW RULES enforced by the implementation:
 *
 *  createBooking  → status starts as PENDING; no conflict check here
 *  approveBooking → PENDING → APPROVED; conflict detection runs HERE
 *  rejectBooking  → PENDING → REJECTED; adminNote AND rejectionReason stored
 *  cancelBooking  → PENDING|APPROVED → CANCELLED
 *                   (users can cancel own; admins can cancel any)
 *  deleteBooking  → SOFT DELETE: sets status = DELETED, records deletedAt
 *                   ADMIN only; preserves audit trail
 *
 * READ METHODS:
 *  getAllBookings     → ADMIN only; supports optional filters
 *  getUserBookings   → returns only the calling user's non-deleted bookings
 *  getBookingById    → anyone; validates existence
 * ================================================================
 */
public interface BookingService {

    /**
     * Create a new booking (any authenticated user).
     * Validates: startTime < endTime, attendees ≥ 1, resource exists & available.
     * Status: PENDING
     */
    BookingDTO.BookingResponse createBooking(String userId, BookingDTO.BookingRequest request);

    /**
     * Approve a PENDING booking (ADMIN only).
     * Runs conflict detection — throws BookingException (409) if slot is taken.
     * Transition: PENDING → APPROVED
     */
    BookingDTO.BookingResponse approveBooking(String bookingId, String adminNote, String adminRole);

    /**
     * Reject a PENDING booking (ADMIN only).
     * Stores both adminNote (general) and rejectionReason (displayed to user).
     * Transition: PENDING → REJECTED
     */
    BookingDTO.BookingResponse rejectBooking(String bookingId, String adminNote, String adminRole);

    /**
     * Cancel a booking.
     *  - USER can cancel their own PENDING or APPROVED bookings.
     *  - ADMIN can cancel any booking.
     * Transition: PENDING|APPROVED → CANCELLED
     */
    BookingDTO.BookingResponse cancelBooking(String bookingId, String requestingUserId, String requestingRole);

    /**
     * Soft-delete a booking (ADMIN only).
     *
     * WHY SOFT DELETE?
     *  Instead of physically removing the document, we set status = DELETED
     *  and record deletedAt. This preserves the audit trail for:
     *   - Billing disputes
     *   - Abuse investigations
     *   - Accidental deletion recovery
     *  Hard deletes are irreversible and break historical reporting.
     *
     * Transition: Any status → DELETED
     * HTTP: Returns 200 with final booking state (or 204 with empty body).
     */
    BookingDTO.BookingResponse deleteBooking(String bookingId, String requestingRole);

    /**
     * Get ALL bookings (ADMIN only) with optional filtering.
     *
     * Filters supported (all optional, can be combined):
     *  - status     : e.g. PENDING, APPROVED, REJECTED
     *  - resourceId : bookings for a specific resource
     *  - userId     : bookings created by a specific user
     *  - date       : bookings whose startTime falls on this LocalDate
     *
     * Filtering logic is applied in the service layer using repository
     * derived queries. DELETED bookings are excluded by default.
     */
    List<BookingDTO.BookingResponse> getAllBookings(BookingDTO.BookingFilter filter);

    /**
     * Get bookings for the currently authenticated user.
     * Excludes soft-deleted bookings.
     */
    List<BookingDTO.BookingResponse> getUserBookings(String userId);

    /** Get a single booking by ID. Throws 404 if not found. */
    BookingDTO.BookingResponse getBookingById(String bookingId);
}

File: frontend/src/services/bookingService.js
/**
 * ================================================================
 * bookingService.js — Booking Management API Layer
 * ================================================================
 * Wraps every backend endpoint defined in BookingController.java.
 *
 * Endpoint map (from BookingController):
 *   POST   /api/bookings                  → createBooking()
 *   GET    /api/bookings?status&resourceId&userId&date → getAllBookings(filters)
 *   GET    /api/bookings/my               → getMyBookings()
 *   GET    /api/bookings/:id              → getById()
 *   PUT    /api/bookings/:id/approve      → approveBooking()
 *   PUT    /api/bookings/:id/reject       → rejectBooking(id, note)
 *   PUT    /api/bookings/:id/cancel       → cancelBooking()
 *   DELETE /api/bookings/:id              → deleteBooking()
 * ================================================================
 */
import api from './axiosConfig'

const BASE = '/api/bookings'

export const bookingService = {
  // ── CREATE ────────────────────────────────────────────────────
  /**
   * Create a new PENDING booking.
   * Body: { resourceId, startTime, endTime, purpose, attendeeCount }
   * startTime/endTime must be ISO-8601 LocalDateTime strings.
   */
  createBooking: (data) =>
    api.post(BASE, data).then(r => r.data),

  // ── READ (ADMIN — with optional filters) ──────────────────────
  /**
   * Get all bookings (ADMIN only).
   * Filters: { status, resourceId, userId, date }  — all optional.
   * date must be "YYYY-MM-DD" string when provided.
   */
  getAllBookings: (filters = {}) => {
    // Build query string from non-null filter values
    const params = new URLSearchParams()
    if (filters.status)     params.append('status',     filters.status)
    if (filters.resourceId) params.append('resourceId', filters.resourceId)
    if (filters.userId)     params.append('userId',     filters.userId)
    if (filters.date)       params.append('date',       filters.date)
    return api.get(`${BASE}?${params.toString()}`).then(r => r.data)
  },

  // ── READ (USER — own bookings) ────────────────────────────────
  /**
   * Get the authenticated user's own bookings.
   * Backend resolves userId from JWT — no param needed.
   */
  getMyBookings: () =>
    api.get(`${BASE}/my`).then(r => r.data),

  /** Get a single booking by ID */
  getById: (id) =>
    api.get(`${BASE}/${id}`).then(r => r.data),

  // ── WORKFLOW ACTIONS ──────────────────────────────────────────
  /**
   * Approve a PENDING booking (ADMIN only).
   * Runs conflict detection on the backend.
   * note is optional admin comment.
   */
  approveBooking: (id, note = '') =>
    api.put(`${BASE}/${id}/approve`, { note }).then(r => r.data),

  /**
   * Reject a PENDING booking (ADMIN only).
   * note is the rejection reason shown to the user.
   */
  rejectBooking: (id, note = '') =>
    api.put(`${BASE}/${id}/reject`, { note }).then(r => r.data),

  /**
   * Cancel a booking.
   * Owners can cancel PENDING or APPROVED bookings.
   * ADMIN can cancel any non-terminal booking.
   */
  cancelBooking: (id) =>
    api.put(`${BASE}/${id}/cancel`).then(r => r.data),

  // ── DELETE (ADMIN — soft-delete) ──────────────────────────────
  /**
   * Soft-delete a booking (ADMIN only).
   * Sets status=DELETED + records deletedAt; document stays in DB.
   */
  deleteBooking: (id) =>
    api.delete(`${BASE}/${id}`).then(r => r.data),
}
